docs/architecture/legacy-module-integration.md

```diff
@@ -0,0 +1,239 @@
+
+    classDef legacy fill:#f0f0f0,stroke:#999999
+    classDef app_common fill:#d5f5d5,stroke:#00aa00
+    classDef new_modules fill:#d0e0ff,stroke:#0066cc
+
+    class APP_COMMON app_common
+    class OTHER_MODULES,INTERFACES,IMPLEMENTATIONS new_modules
+    class LEGACY1,LEGACY2,LEGACY3 legacy
+```
+
+### Implementation Techniques
+
+Several techniques are used to implement the bridge pattern effectively:
+
+1. **Wrapper Classes**: Creating immutable data classes that wrap legacy data structures, implementing interfaces from the new architecture. These wrappers typically include conversion methods to transform between legacy and new data structures.
+
+2. **Adapter Implementations**: Classes in `:app-common` that implement interfaces from the new architecture but delegate to legacy code internally.
+
+3. **Data Conversion**: Static conversion methods (often in companion objects) that handle mapping between legacy and new data structures, ensuring clean separation of concerns.
+
+#### Example: Account Profile Bridge
+
+A concrete example of this pattern is the account profile bridge, which demonstrates a complete implementation of the bridge pattern across multiple layers:
+
+1. **Modern Interfaces**:
+   - `AccountProfileRepository` in `feature:account:api` defines the high-level contract for account profile management
+   - `AccountProfileLocalDataSource` in `feature:account:core` defines the data access contract
+2. **Modern Data Structure**: `AccountProfile` in `feature:account:api` is a clean, immutable data class that represents account profile information in the new architecture.
+3. **Repository Implementation**: `DefaultAccountProfileRepository` in `feature:account:core` implements the `AccountProfileRepository` interface and depends on `AccountProfileLocalDataSource`.
+4. **Bridge Implementation**: `DefaultAccountProfileLocalDataSource` in `app-common` implements the `AccountProfileLocalDataSource` interface and serves as the bridge to legacy code.
+5. **Legacy Access**: The bridge uses `LegacyAccountWrapperManager` to access legacy account data:
+   - `LegacyAccountWrapperManager` in `core:android:account` defines the contract for legacy account access
+   - `LegacyAccountWrapper` in `core:android:account` is an immutable wrapper around the legacy `LegacyAccount` class
+6. **Data Conversion**: The bridge converts between modern `AccountProfile` objects and legacy account data via `LegacyAccountWrapper`.
+7. **Dependency Injection**: The `appCommonAccountModule` in `app-common` registers `DefaultAccountProfileLocalDataSource` as implementations of the respective interface.
+
+This multi-layered approach allows newer modules to interact with legacy account functionality through clean, modern interfaces without directly depending on legacy code. It also demonstrates how bridges can be composed, with higher-level bridges (AccountProfile) building on lower-level bridges (LegacyAccountWrapper).
+
+## Testing Considerations
```

Comment:
Some more examples of well written tests for this would be great to point to.

Reply:
This is the hard part, hence there are no good examples. It mostly depends on which legacy component you're trying to wrap. So they are tailor fit for the problem at hand and might even skip testing some aspects if the involved legacy code is not testable (meaning high effort and time consuming).